### PR TITLE
Typescript express stack

### PR DESCRIPTION
**Pull request by SaifRehman**
### Checklist:

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stacks-overview.md#stack-structure).

### Modifying an existing stack:

- [x] Updated the stack version in `stack.yaml`

### Contributing a new stack:

- Describe how application dependencies are managed:

- Explain how Appsody file watcher is utilized:

- Describe other Appsody environment variables defined by the stack image:

- Describe any limitations and known issues:

### CODE REVIEW

**Comment by SaifRehman** [comment]
@claassistantio Thank you for considering. I have signed the CLA.

**Comment by sam-github** [comment]
I'm not a typescript user, but it was my understanding that if the app was written in typescript, it could be mounted into the nodejs-express app, does that not work? The top-level app doesn't have to be written in ts itself in order to use a ts-implemented and exported app.

Alternatively, is it possible to inherit from the nodejs-express base stack image, adding only the differences?

I'm somewhat concerned about the maintainability of this, it contains a lot of copy-n-paste from nodejs-express, particularly the README, and a whole bunch of typescript boilerplate that I don't understand. Are you comitting to supporting this going forward and copying the changes over?

I'd be remiss for not mentioning that loopback was written in typescript.

cc: @raymondfeng

**Comment by SaifRehman** [comment]
Hi @sam-github , thank you for your feedback. I created this stack because all my backend microservices was creating using typescript express stack I built. I was actually trying to build typescript-express template along with mongodb. However, when I had a conversation with appsody team, they told they need one typescript stack. So I decided to create one and make a PR. Here are my following answers to your questions 

1) According to my understanding you cant mount typescript to nodejs express app. You actually need a transpiler to transpile typescript code to javascript, this is achieved by uses babel along with webpack and typescript transpiler. Typescript is not natively supported by node runtime, but there is something called Demo which is node runtime for typescript. Hence you do need custom template for typescript, which I have created.

2. Yes I will be maintaining the project 

3. Yes, most of readme parts are same form node-express stack. Because it is almost similar, the only difference is that, it is built using typescript. Along with that we have webpack and babel integration.

4. Loopback template is different from this. This is a generic typescript template developer can use to build microservices using typescript

What part of this typescript template you don't understand?. Please elaborate on that.

Thanks

**Comment by sam-github** [comment]
Specifically, I do not understand why typescript cannot be handled as another template in https://github.com/appsody/stacks/tree/master/incubator/nodejs-express/templates Why can't the app's transpilation dependencies and setup be put into a typescript specific version of https://github.com/appsody/stacks/blob/master/incubator/nodejs-express/templates/scaffold/package.json ? Did you try? What specific issues did you encounter?

Transpilation has to occur on-change (so modified code is transpiled). That appears to me to require either a modified start script in the ts  version of scaffold/package.json (should work fine), or a derived stack with one small change to 
https://github.com/appsody/stacks/blob/master/incubator/nodejs-express/image/Dockerfile-stack#L14

**Comment by Kamran64** [comment]
@SaifRehman - Thanks for contributing this. Please can you take a look at the travis build errors. Looks like it's failing on the `appsody test` and `appsody build` portions. I notice there's no `APPSODY_TEST` env var set so I'm curious as to why/where `npm test` is being called (the cause of the failure).

File: experimental/typescript-express/stack.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+name: Typescript template
+version: 0.1.1
```
Comment (Kamran64) [inline experimental/typescript-express/stack.yaml]: This should be 0.1.0 for an initial version

File: experimental/typescript-express/image/project/package.json [hunk 2]
```diff
@@ -0,0 +1,84 @@
+{
+  "name": "express-react-typescript",
```
Comment (Kamran64) [inline experimental/typescript-express/image/project/package.json]: Should this field have the same name as the stack id?

File: experimental/typescript-express/image/project/package.json [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  "version": "1.0.0",
```
Comment (Kamran64) [inline experimental/typescript-express/image/project/package.json]: 0.1.0?

**Comment by neeraj-laad** [comment]
@SaifRehman just wondered if you have looked at previous comments. We've had people ask for a typescript stack recently and thought it would be a good addition as long as we know it should be a separate stack and not part of an existing stack and it passes our stack criteria.

**Comment by Kamran64** [comment]
@SaifRehman have you had a chance to look at this yet?

**Comment by SaifRehman** [comment]
Hi Team,
I will get back on this topic. I had been busy with other work recently

**Comment by Kamran64** [comment]
Summarising the conversation I had with @SaifRehman here:
- @SaifRehman will add test cases to resolve `appsody test` issues and investigate the problems with `appsody build` (although `appsody build` passes locally)
- This stack will be kept as a separate new stack and not a template to `nodejs-express` supposedly due to a lot of extra configuration/dependencies required. I think this is okay as it's being added to experimental too.

**Comment by neeraj-laad** [comment]
@SaifRehman Are you able to address review comments mentioned above?

**Comment by neeraj-laad** [comment]
Closing this due to inactivity. Please re-open when the review comments have been addressed.

**Pull request closed without merging**